packages/provider/src/tasks/calculateSolutions.ts

```diff
@@ -0,0 +1,84 @@
+import { CaptchaStates } from '@prosopo/types'
+import { ProsopoEnvError, logger } from '@prosopo/common'
+import { ProviderEnvironment } from '@prosopo/types-env'
+import { ScheduledTaskNames } from '@prosopo/types'
+import { Tasks } from './tasks'
+import { calculateNewSolutions, checkIfTaskIsRunning, updateSolutions } from '../util'
+import { captchaSort } from '@prosopo/datasets'
+export class CalculateSolutionsTask extends Tasks {
+    constructor(env: ProviderEnvironment) {
+        super(env)
+        this.logger = logger(env.config.logLevel, 'CalculateSolutionsTask')
+    }
+
+    /**
+     * Apply new captcha solutions to captcha dataset and recalculate merkle tree
+     */
+    async run(): Promise<number> {
+        try {
+            const taskRunning = await checkIfTaskIsRunning(ScheduledTaskNames.CalculateSolution, this.db)
+            if (!taskRunning) {
+                // Get the current datasetId from the contract
+                const provider = (await this.contract.methods.getProvider(this.contract.pair.address, {})).value
+                    .unwrap()
+                    .unwrap()
+
+                // Get any unsolved CAPTCHA challenges from the database for this datasetId
+                const unsolvedCaptchas = await this.db.getAllCaptchasByDatasetId(
+                    provider.datasetId.toString(),
+                    CaptchaStates.Unsolved
+                )
+
+                // edge case when a captcha dataset contains no unsolved CAPTCHA challenges
```

Comment:
Redundant check since additionally checked on line 45

Reply:
Nice spot, removed.

Reply:
I removed the check but it's required by TypeScript...